tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/UsageReport/WritePackageUsageData.cs

```diff
@@ -180,28 +180,44 @@ public override bool Execute()
                 assetFiles,
                 assetFile =>
                 {
-                    var properties = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                    JObject jObj;
 
                     using (var file = File.OpenRead(Path.Combine(RootDir, assetFile)))
                     using (var reader = new StreamReader(file))
                     using (var jsonReader = new JsonTextReader(reader))
                     {
-                        while (jsonReader.Read())
-                        {
-                            if (jsonReader.TokenType == JsonToken.PropertyName &&
-                                jsonReader.Value is string value)
-                            {
-                                properties.Add(value);
-                            }
-                        }
+                        jObj = (JObject)JToken.ReadFrom(jsonReader);
                     }
 
+                    var properties = new HashSet<string>(
+                        jObj.SelectTokens("$.targets.*")?.Children()
+                            .Concat(jObj.SelectToken("$.libraries"))
+                            .Select(t => ((JProperty)t).Name)
+                            .Distinct(), 
+                        StringComparer.OrdinalIgnoreCase);
+
+                    var directDependencies = jObj.SelectTokens("$.project.frameworks.*.dependencies")?.Children().Select(dep =>
+                        new
+                        {
+                            name = ((Newtonsoft.Json.Linq.JProperty)dep).Name,
+                            target = dep.SelectToken("$..target")?.ToString(),
+                            version = VersionRange.Parse(dep.SelectToken("$..version")?.ToString()), //dep.SelectToken("$..version")?.ToString().Replace("[", "").Replace(", )", ""),
```

Comment:
More `?.`s on `SelectToken`, but `VersionRange.Parse` throws on null anyway.

Reply:
It's required on SelectToken, because it can return null and I need the ToString() here.

Reply:
Ah, didn't notice the plural difference. However, why is it better to wait to throw an exception in `VersionRange.Parse` than throw it on `ToString`?

Reply:
I'm not sure it's better one way or the other.  Either way, it will fail if we have a dependency that does not have a version.  In the case of `VersionRange.Parse` throwing, we will know it failed trying to parse a version, which may be a quicker indication of the issue, but in the end, it is going to fail.

Reply:
To make it extra diagnosable, you can try this syntax, not sure which C# version it was added in though:

```C#
dep.SelectToken("$..version")?.ToString()
?? throw new NotSupportedException($"Direct dependency has no version in '{assetFile}': '{dep}'")
```

I agree that in `?.` vs. `.`, `?.` is more diagnosable in this case, but it's less readable because it's easy to wonder why it's `?.` rather than `.`.

Reply:
I would argue that it is more readable because it matches all of the other attribute accessors, which 1) select a token, 2) check for null and 3) call `ToString`.

Reply:
What makes it less readable to me is that when I see it, I stop and wonder "is this a bug?" because it appears that `?.` is *intended* to make the code robust to `null`s even though it doesn't.

In any case... adding the exception throw expression would make it clear on first reading, clarify the runtime error way more (probably even diagnosable from CI artifacts), and keep `?.` self-similarity.

This is really minor and I don't mind merging anyway.